src/compiler/checker.ts

```diff
@@ -12419,11 +12414,14 @@ namespace ts {
             }
             if ((<Identifier>node.name).text === "this") {
                 if (indexOf(func.parameters, node) !== 0) {
-                    error(node, Diagnostics.this_parameter_must_be_the_first_parameter);
+                    error(node, Diagnostics.A_this_parameter_must_be_the_first_parameter);
                 }
-                if (func.kind === SyntaxKind.Constructor) {
+                if (func.kind === SyntaxKind.Constructor || func.kind === SyntaxKind.ConstructSignature) {
                     error(node, Diagnostics.A_constructor_cannot_have_a_this_parameter);
                 }
+                if (func.kind === SyntaxKind.SetAccessor) {

```

Comment:
Are you looking at the latest commits? I added to the check on line 12419 and an associated test (with an interface, not a variable, but they fail the same way) in thisTypeInFunctionsNegative.ts